Revise data source consolidation for `ListStakePools`.

jonathanknowles opened the pull request:
# Issue Number

#1930 
#1948 

# Overview

The `ListStakePools` endpoint produces a list of `ApiStakePool` records by merging data from (at least) **two sources**:

- DB data: data fetched from the local pool database. The presence of DB data for a pool indicates that the wallet has seen one or more registration certificates for that pool. 
- LSQ data: pool metrics data fetched with local state query. 

This PR makes **two changes** to the set of pools returend by the `ListStakePools` endpoint:

| DB data available? | LSQ data available? | Old Behaviour | New Behaviour |
|---|---|---|---|
| | | | |
| :heavy_check_mark: | :negative_squared_cross_mark: | pool **not** listed | pool **listed** |
| | | | stake pool metrics are assigned a default value of zero |
| | | | |
| :negative_squared_cross_mark:  | :heavy_check_mark: | pool **listed** | pool **not** listed |
| | | the following fields have missing values:<br> - `cost` <br> - `margin` <br> - `pledge` |  the following fields are now required:<br> - `cost` <br> - `margin` <br> - `pledge` |

jonathanknowles committed:
Use `db` instead of `chain` within `combineDbAndLsqData`.

Justification: consistency of value names with types.

jonathanknowles committed:
Reformat code to comply with coding standards.

jonathanknowles committed:
Report all pool metric values as zero when real pool LSQ data is absent.

When a pool is registered (with a registration certificate) but not yet
active (and therefore not yet causing pool metrics data to be available
over local state query), we would nevertheless still prefer to include
that pool in the list returned by the `ListStakePools` API endpoint.

This change causes such pools to be listed, and reports all pool metric
values as **zero** when pool LSQ data is absent.

KtorZ commented on lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs:
```diff
@@ -198,15 +198,27 @@ combineDbAndLsqData
+        pure $ mkApiPool k lsqDefault $ Just db
+      where
+        lsqDefault = PoolLsqData
+            { nonMyopicMemberRewards = minBound
```
This needs to be checked / justified against the specs. I am not sure that `0` would be a significant value for new pools.

KtorZ replied:
@jonathanknowles after checking, the ledger team suggests that (given the time constraints) we use the average non myopic member rewards of the top `k` pools (where `k` is the desired number of pools).

Anviking commented on lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs:
```diff
@@ -208,15 +208,27 @@ combineDbAndLsqData
-    -- In case our chain following has missed a retirement certificate, we
-    -- treat the lsq data as the source of truth, and dropMissing here.
-    chainButNoLsq = dropMissing
+    lsqButNoDb = traverseMissing $ \k lsq ->
```
Should we not rely 100% on the DB now, and `dropMissing` on `lsqButNoDb`?

jonathanknowles replied:
> Should we not rely 100% on the DB now, and `dropMissing` on `lsqButNoDb`?

We could indeed do that. If we did, we'd be able to guarantee the presence of the following fields for each `ApiStakePool` object we produce:

- `cost`
- `margin`
- `pledge`

(See: https://github.com/input-output-hk/cardano-wallet/blob/master/lib/core/src/Cardano/Wallet/Api/Types.hs#L406-L408)

We'd then be able to make the following simplification:
```patch
@@ -403,9 +403,9 @@ data ApiStakePool = ApiStakePool
     { id :: !(ApiT PoolId)
     , metrics :: !ApiStakePoolMetrics
     , metadata :: !(Maybe (ApiT StakePoolMetadata))
-    , cost :: !(Maybe (Quantity "lovelace" Natural))
-    , margin :: !(Maybe (Quantity "percent" Percentage))
-    , pledge :: !(Maybe (Quantity "lovelace" Natural))
+    , cost :: !(Quantity "lovelace" Natural)
+    , margin :: !(Quantity "percent" Percentage)
+    , pledge :: !(Quantity "lovelace" Natural)
     , retirement :: !(Maybe ApiEpochInfo)
     } deriving (Eq, Generic, Show)
```

jonathanknowles replied:
> Should we not rely 100% on the DB now, and dropMissing on lsqButNoDb?

Fixed in 457ad036c1e0f7f93e637c8453e6ec5f49c9153c.

jonathanknowles replied:
> We'd then be able to make the following simplification:
> ```patch
> @@ -403,9 +403,9 @@ data ApiStakePool = ApiStakePool
>      { id :: !(ApiT PoolId)
>      , metrics :: !ApiStakePoolMetrics
>      , metadata :: !(Maybe (ApiT StakePoolMetadata))
> -    , cost :: !(Maybe (Quantity "lovelace" Natural))
> -    , margin :: !(Maybe (Quantity "percent" Percentage))
> -    , pledge :: !(Maybe (Quantity "lovelace" Natural))
> +    , cost :: !(Quantity "lovelace" Natural)
> +    , margin :: !(Quantity "percent" Percentage)
> +    , pledge :: !(Quantity "lovelace" Natural)
>      , retirement :: !(Maybe ApiEpochInfo)
>     } deriving (Eq, Generic, Show)
> ```

Fixed in:
- efbfd59beee7bc3feac20c0d3d365c073a5c45b4 (Simplify field types for `ApiStakePool`)
- 4a0274329190c6c8fba9a64a2d063028c3782c14 (Mark fields as required in specification for `ApiStakePool`)
- bec56c630fa48a0614b39f8ce2b8b56ed146eb46 (Regenerate golden test data for `ApiStakePool`)

jonathanknowles committed:
Do not create an `ApiStakePool` if there is no known registration certificate.

In response to review feedback here:

https://github.com/input-output-hk/cardano-wallet/pull/1936#discussion_r460448980

jonathanknowles committed:
Simplify field types for `ApiStakePool`.

The following fields for `ApiStakePool` are always present:

- `cost`
- `margin`
- `pledge`

Therefore, we can simplify the type, removing the `Maybe` wrappers.

jonathanknowles committed:
Simplify assignment to `ApiStakePool` `metadata` field.

jonathanknowles committed:
Apply eta reduction to function `bothPresent` in `combineDbAndLsqData`.

jonathanknowles committed:
Mark fields as required in specification for `ApiStakePool`.

jonathanknowles committed:
Regenerate golden test data for `ApiStakePool`.

jonathanknowles committed:
Change `knownPools` to return the list of pools for which we have certificates.

This change makes `knownPools` consistent with `listPools`.

If a pool appears in `ListStakePools`, it should be possible to
join that pool with `JoinStakePool`.

In response to review feedback:

https://github.com/input-output-hk/cardano-wallet/pull/1936#issuecomment-664206113
https://github.com/input-output-hk/cardano-wallet/pull/1936#issuecomment-664215013

jonathanknowles committed:
Do not convert the set of known pools into a list.

If the sole purpose of `knownPools` is to provide a set of pools that
`joinStakePool` can use in a membership check, then it makes no sense
to convert this set into a list.
